src/job.rs

```diff
@@ -253,285 +252,32 @@ impl Job {
-        }
-    }
-
-    #[test]
-    fn push_pop_back() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.pop_back();
-        }
-        assert_eq!(queue.sentinel, queue.tail);
-    }
-
-    #[test]
-    fn push2_pop2_back() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-        let job2 = Job::from_usize(&2);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.push_back(&job2);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 2);
-
-        unsafe {
-            queue.pop_back();
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.pop_back();
-        }
-        assert_eq!(queue.sentinel, queue.tail);
-    }
-
-    #[test]
-    fn push_pop_front() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        let job = unsafe { queue.pop_front().unwrap() };
-        assert_eq!(job.as_usize(), 1);
-        assert!(job.prev.get().is_none());
-        assert!(job.fut_or_next.get().is_some());
-
-        unsafe {
-            job.drop();
-        }
-
-        assert_eq!(queue.sentinel, queue.tail);
-    }
-
-    #[test]
-    fn push2_pop2_front() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-        let job2 = Job::from_usize(&2);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.push_back(&job2);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 2);
-
-        let job = unsafe { queue.pop_front().unwrap() };
-        assert_eq!(job.as_usize(), 1);
-        assert!(job.prev.get().is_none());
-        assert!(job.fut_or_next.get().is_some());
-
-        unsafe {
-            job.drop();
-        }
-
-        let job = unsafe { queue.pop_front().unwrap() };
-        assert_eq!(job.as_usize(), 2);
-        assert!(job.prev.get().is_none());
-        assert!(job.fut_or_next.get().is_some());
-
-        unsafe {
-            job.drop();
-        }
-
-        assert_eq!(queue.sentinel, queue.tail);
+        // `Job` is still alive as per contract in `push_back`
+        let job = unsafe { val.as_ref() };
```

Comment:
Nit: No need for an extra variable.

```suggestion
        let job = unsafe { self.0.pop_front()?.as_ref() };
```